src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -11,84 +11,200 @@
+    // the middle of the term.
+    auto committed_offset = _raft->committed_offset();
+    if (committed_offset > last_applied()) {
+        // The STM is catching up.
+        auto wait_res = co_await wait_no_throw(
+          committed_offset, ss::lowres_clock::now() + ctp_stm_sync_timeout, as);
+        if (!wait_res) {
+            vlog(
+              _log.warn,
+              "Failed to wait for committed offset {} in term {}",
+              committed_offset,
+              _raft->term());
+            co_return false;
+        }
+    }
+    // NOTE: there could be an in-flight replication that will be committed
+    // after the wait_no_throw() and the state update will be scheduled right
+    // after this method returns. This is fine because the state can tolerate
+    // races.
+    co_return true;
+}
+
+ss::future<std::optional<cluster_epoch>> ctp_stm::get_inactive_epoch() {
+    // Consume the first epoch from the partition starting from
+    // start offset if nothing was reconciled yet or from the last
+    // reconciled offset + 1 otherwise.
+    auto so = _raft->start_offset();
+    auto co = _raft->committed_offset();
+
+    // NOTE: it's enough to read log starting from the start offset
+    // for correctness. However, the local retention could be updated
+    // with some arbitrary lag. In order to avoid holding data for too
+    // long we're trying to read starting from the LRO.
+    auto lro = _state.get_last_reconciled_log_offset().value_or(
+      model::prev_offset(so));
+
+    // NOTE: we can't make a decision to skip the log read based on offsets
+    // because the LRO contains the translated LRO value. But in order for the
+    // LRO to be applied the command should be replicated first. So naturally
+    // this command will be the last in the log and this will make the committed
+    // offset to be greater than the LRO log translated offset.
 
-    // Note: do_apply will be called multiple times with the same batch if it
-    // throws. The method must be idempotent. Because of this we use the batch
-    // base offset as dl_version rather than record offset. This also means that
-    // in the same batch we cannot add an overlay and then remove it. Other
-    // caveats may exist as well.
-    //
-    // The version can't go backwards but in case of a partial apply and a retry
-    // it could.
-    auto new_dl_version = dl_version(batch.base_offset());
-
-    if (!_state.get_offsets().can_apply(new_dl_version)) {
+    storage::log_reader_config cfg(
+      model::next_offset(lro),
+      co,
+      0,
+      4_MiB,
+      std::make_optional(model::record_batch_type::dl_placeholder),
+      std::nullopt,
+      std::nullopt);
+
+    auto reader = co_await _raft->make_reader(cfg);
+    auto result = co_await std::move(reader).consume(
+      ctp_stm_consumer{}, model::no_timeout);
+    if (result.has_value()) {
+        auto epoch = result.value();
+        auto inactive = prev_cluster_epoch(epoch);
```

Comment:
we should make it a goal to get rid of the function `prev_cluster_epoch`. it's strange to have it in a system where the operations are epoch++ and epoch0 <[=] epoch1. this might be as simple as changing the definition of a return value such that the caller users `<` instead of `<=` or something like that.